### PR TITLE
First release of E2E

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
This is the first release to support end to end encryption.
Server app is needed: https://github.com/nextcloud/end_to_end_encryption

### CODE REVIEW

File: src/com/owncloud/android/lib/common/network/WebdavEntry.java [hunk 1]
```diff
@@ -210,6 +213,19 @@ public WebdavEntry(MultiStatusResponse ms, String splitElement) {
                 mIsFavorite = false;
             }
 
+            // NC encrypted property <nc:is-encrypted>
+            prop = propSet.get(EXTENDED_PROPERTY_IS_ENCRYPTED,  Namespace.getNamespace(NAMESPACE_NC));
+            if (prop != null) {
+                String encryptedValue = (String) prop.getValue();
+                if ("1".equals(encryptedValue)) {
```
Comment (AndyScherzinger) [inline src/com/owncloud/android/lib/common/network/WebdavEntry.java]: Can we make ```"1"``` a constant so this gets the look of a business rule to be read like ```if (IS_ENCRYPTED.equals(encryptedValue))``` ?

File: src/com/owncloud/android/lib/resources/files/GetMetadataOperation.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                ArrayList<Object> metadataArray = new ArrayList<>();
+                metadataArray.add(metadata);
+                result.setData(metadataArray);
+            } else {
+                result = new RemoteOperationResult(false, getMethod);
+                client.exhaustResponse(getMethod.getResponseBodyAsStream());
+            }
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (AndyScherzinger) [inline src/com/owncloud/android/lib/resources/files/GetMetadataOperation.java]: Please use a logger instead :)

Comment (AndyScherzinger) [reply src/com/owncloud/android/lib/resources/files/GetMetadataOperation.java]: nevermind, it is just in the next line.

**Comment by AndyScherzinger** [comment]
Only minor comments, only thing I am unsure about is if the ```e.printStackTrace();``` statements should be removed.

File: src/com/owncloud/android/lib/resources/files/GetMetadataOperation.java (hunk 2 above)
Comment (tobiasKaminsky) [reply src/com/owncloud/android/lib/resources/files/GetMetadataOperation.java]: 👍 still it is not useful to print a stacktrace.

File: src/com/owncloud/android/lib/common/OwnCloudClient.java [hunk 3]
```diff
@@ -71,31 +71,36 @@
...
+            userAgent = OwnCloudClientManagerFactory.getUserAgent();
+        }
         
-        getParams().setCookiePolicy(
-        		CookiePolicy.IGNORE_COOKIES);
-        getParams().setParameter(
-        		PARAM_SINGLE_COOKIE_HEADER, 			// to avoid problems with some web servers
-        		PARAM_SINGLE_COOKIE_HEADER_VALUE);
+        getParams().setParameter(HttpMethodParams.USER_AGENT, userAgent);
+        getParams().setParameter(PARAM_PROTOCOL_VERSION, HttpVersion.HTTP_1_1);
```
Comment (mario) [inline src/com/owncloud/android/lib/common/OwnCloudClient.java]: Why 1.1?

File: src/com/owncloud/android/lib/resources/files/LockFileOperation.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                ArrayList<Object> tokenArray = new ArrayList<>();
+                tokenArray.add(token);
+                result.setData(tokenArray);
+            } else {
+                result = new RemoteOperationResult(false, postMethod);
+                client.exhaustResponse(postMethod.getResponseBodyAsStream());
+            }
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/LockFileOperation.java]: Please remove.

File: src/com/owncloud/android/lib/resources/files/UnlockFileOperation.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        RemoteOperationResult result;
+        DeleteMethod deleteMethod = null;
+
+        try {
+            // remote request
+            deleteMethod = new DeleteMethod(client.getBaseUri() + LOCK_FILE_URL + localId);
+            deleteMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
+            deleteMethod.addRequestHeader("Content-Type", "application/x-www-form-urlencoded");
+            deleteMethod.addRequestHeader("token", token);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/UnlockFileOperation.java]: put "token" in  private static final variable please. Not a huge fan of this, but consistency *shrug*

File: src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        PutMethod putMethod = null;
+        RemoteOperationResult result;
+
+        try {
+            // remote request
+            putMethod = new PutMethod(client.getBaseUri() + METADATA_URL + fileId);
+            putMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
+            putMethod.addRequestHeader("Content-Type", "application/x-www-form-urlencoded");
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java]: Content-Type as private static final pls.

File: src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+            NameValuePair[] putParams = new NameValuePair[2];
+            putParams[0] = new NameValuePair("token", token);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java]: Token and  format too pls.

File: src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                ArrayList<Object> keys = new ArrayList<>();
+                keys.add(metadata);
+                result.setData(keys);
+            } else {
+                result = new RemoteOperationResult(false, putMethod);
+                client.exhaustResponse(putMethod.getResponseBodyAsStream());
+            }
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/UpdateMetadataOperation.java]: Remove.

File: src/com/owncloud/android/lib/resources/users/GetPrivateKeyOperation.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                keys.add(key);
+                result.setData(keys);
+            } else {
+                result = new RemoteOperationResult(false, getMethod);
+                client.exhaustResponse(getMethod.getResponseBodyAsStream());
+            }
+
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/users/GetPrivateKeyOperation.java]: Remove.

File: src/com/owncloud/android/lib/resources/users/GetPublicKeyOperation.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                ArrayList<Object> keys = new ArrayList<>();
+                keys.add(key);
+                result.setData(keys);
+            } else {
+                result = new RemoteOperationResult(false, getMethod);
+                client.exhaustResponse(getMethod.getResponseBodyAsStream());
+            }
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/users/GetPublicKeyOperation.java]: Remove.

File: src/com/owncloud/android/lib/resources/users/SendCSROperation.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        PostMethod postMethod = null;
+        RemoteOperationResult result;
+
+        try {
+            // remote request
+            postMethod = new PostMethod(client.getBaseUri() + PUBLIC_KEY_URL + JSON_FORMAT);
+            postMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
+            postMethod.setParameter("csr", csr);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/users/SendCSROperation.java]: csr s private static final pls.

File: src/com/owncloud/android/lib/resources/users/SendCSROperation.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                keys.add(key);
+                result.setData(keys);
+            } else {
+                result = new RemoteOperationResult(false, postMethod);
+                client.exhaustResponse(postMethod.getResponseBodyAsStream());
+            }
+
+        } catch (Exception e) {
+            result = new RemoteOperationResult(e);
+            e.printStackTrace();
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/users/SendCSROperation.java]: Remove.

File: src/com/owncloud/android/lib/resources/users/StorePrivateKeyOperation.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        PostMethod postMethod = null;
+        RemoteOperationResult result;
+
+        try {
+            // remote request
+            postMethod = new PostMethod(client.getBaseUri() + PRIVATE_KEY_URL + JSON_FORMAT);
+            postMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
+            postMethod.setParameter("privateKey", privateKey);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/users/StorePrivateKeyOperation.java]: privateKey as private static final.

File: src/com/owncloud/android/lib/common/OwnCloudClient.java (hunk 3 above)
Comment (tobiasKaminsky) [reply src/com/owncloud/android/lib/common/OwnCloudClient.java]: I have no idea ;-)
It is on master all the time: https://github.com/nextcloud/android-library/blob/master/src/com/owncloud/android/lib/common/OwnCloudClient.java#L92

**Comment by tobiasKaminsky** [comment]
@mario thanks for the CR, I fixed everything and rebased the PR.

File: src/com/owncloud/android/lib/resources/files/LockFileOperation.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        RemoteOperationResult result;
+        PostMethod postMethod = null;
+
+        try {
+            postMethod = new PostMethod(client.getBaseUri() + LOCK_FILE_URL + localId + JSON_FORMAT);
+
+            if (!token.isEmpty()) {
+                postMethod.setParameter("token", token);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/LockFileOperation.java]: token should be private static final.

File: src/com/owncloud/android/lib/resources/files/StoreMetadataOperation.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    protected RemoteOperationResult run(OwnCloudClient client) {
+        PostMethod postMethod = null;
+        RemoteOperationResult result;
+
+        try {
+            // remote request
+            postMethod = new PostMethod(client.getBaseUri() + METADATA_URL + fileId + JSON_FORMAT);
+            postMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
+            postMethod.setParameter("metaData", encryptedMetadataJson);
```
Comment (mario) [inline src/com/owncloud/android/lib/resources/files/StoreMetadataOperation.java]: metaData should be private static final.

**Comment by mario** [comment]
@tobiasKaminsky after you fix the two remaining things, feel free to merge. Good job! :)

👍

**Comment by tobiasKaminsky** [comment]
Merging with only one approval as @AndyScherzinger is on vacation.

**Pull request merged**